src/exp/istio-watcher/pkg/watcher/watcher.go

```diff
@@ -0,0 +1,291 @@
+	p.DestinationNamespace = strings.Split(p.DestinationNamespace, ".")[1]
+	p.RequestPath = strings.Split(p.RequestPath, ".")[1]
+}
+
+type EnvoyMetrics struct {
+	Stats []Metric `json:"stats"`
+}
+
+type Metric struct {
+	Name string `json:"name"`
+}
+
+func NewWatcher(mapperClient mapperclient2.MapperClient) (*IstioWatcher, error) {
+	//conf, err := rest.InClusterConfig()
+	//if err != nil {
+	//	return nil, err
+	//}
+	conf, err := clientcmd.BuildConfigFromFlags("", filepath.Join(homedir.HomeDir(), ".kube", "config"))
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &IstioWatcher{
+		clientset:    clientset,
+		config:       conf,
+		mapperClient: mapperClient,
+		connections:  map[*ConnectionWithPath]time.Time{},
+	}
+
+	return m, nil
+}
+
+func (m *IstioWatcher) Flush() map[*ConnectionWithPath]time.Time {
+	r := m.connections
+	m.connections = map[*ConnectionWithPath]time.Time{}
+	return r
+}
+
+func (m *IstioWatcher) CollectIstioConnectionMetrics(ctx context.Context, namespace string) error {
+	sendersErrGroup, sendersCtx := errgroup.WithContext(ctx)
+	receiversErrGroup, _ := errgroup.WithContext(ctx)
+	metricsChan := make(chan *EnvoyMetrics, MetricsBufferedChannelSize)
+	done := make(chan int)
+	defer close(done)
+
+	podList, err := m.clientset.CoreV1().Pods(namespace).List(ctx, v1.ListOptions{LabelSelector: IstioPodsLabelSelector})
+	if err != nil {
+		return err
+	}
+
+	for _, pod := range podList.Items {
+		// Known for loop gotcha with goroutines
+		curr := pod
+		sendersErrGroup.Go(func() error {
+			if err := m.getEnvoyMetricsFromSidecar(sendersCtx, curr, metricsChan); err != nil {
+				logrus.Errorf("Failed fetching request metrics from pod %s", curr.Name)
+				return err
+			}
+			return nil
+		})
+	}
+	receiversErrGroup.Go(func() error {
+		// Function call below updates a map which isn't concurrent-safe.
+		// Needs to be taken into consideration if the code should ever change to use multiple goroutines
+		if err := m.convertMetricsToConnections(metricsChan, done); err != nil {
+			return err
+		}
+		return nil
+	})
+
+	if err := sendersErrGroup.Wait(); err != nil {
+		return err
+	}
+	done <- 0
+
+	if err := receiversErrGroup.Wait(); err != nil {
+		return err
+	}
+
+	close(metricsChan)
+	return nil
+}
+
+func (m *IstioWatcher) getEnvoyMetricsFromSidecar(ctx context.Context, pod corev1.Pod, metricsChan chan<- *EnvoyMetrics) error {
+	req := m.clientset.CoreV1().
+		RESTClient().
+		Post().
+		Resource("pods").
+		Name(pod.Name).
+		Namespace(pod.Namespace).SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Command:   strings.Split(IstioProxyTotalRequestsCMD, " "),
+			Stdout:    true, // We omit stderr and we error according to return code from executed cmd
+			Container: IstioSidecarContainerName,
+		}, scheme.ParameterCodec)
+
+	// TODO: use error group context in exec
```

Comment:
do eet
although not sure i understand the todo? where would the ctx go? in StreamWithContext? and what is the errorgroup?

Reply:
TODO was written before I used `StreamWithContext`, but just `Stream` so I figured I'll need to do something with the context.